tools/hula/proxy/src/proxy.rs

```diff
@@ -0,0 +1,222 @@
+use log::{debug, error, info, warn};
+use rmp_serde::{encode::write_named, from_slice};
+
+use crate::{
+    idle::{charging_skull, send_idle},
+    listener::HulaListener,
+};
+
+const HULA_SOCKET_PATH: &str = "/tmp/hula";
+const HULA_TIMEOUT: Duration = Duration::from_secs(1);
+const LOLA_SOCKET_PATH: &str = "/tmp/robocup";
+const LOLA_SOCKET_RETRY_COUNT: usize = 60;
+const LOLA_SOCKET_RETRY_INTERVAL: Duration = Duration::from_secs(1);
+
+fn wait_for_lola() -> Result<UnixStream> {
+    for _ in 0..LOLA_SOCKET_RETRY_COUNT {
+        if let Ok(socket) = UnixStream::connect(LOLA_SOCKET_PATH) {
+            return Ok(socket);
+        }
+        info!("Waiting for LoLA socket to become available...");
+        sleep(LOLA_SOCKET_RETRY_INTERVAL);
+    }
+    bail!("stopped after {} retries", LOLA_SOCKET_RETRY_COUNT)
+}
+
+pub struct Proxy {
+    lola: UnixStream,
+    hula: HulaListener,
+    epoll_fd: RawFd,
+}
+
+impl Proxy {
+    pub fn initialize() -> Result<Self> {
+        let lola = wait_for_lola().wrap_err("failed to connect to LoLA")?;
+        let hula = HulaListener::bind(HULA_SOCKET_PATH)
+            .wrap_err_with(|| format!("failed to bind {}", HULA_SOCKET_PATH))?;
+
+        let epoll_fd = epoll::create(false).wrap_err("failed to create epoll file descriptor")?;
+        add_to_epoll(epoll_fd, lola.as_raw_fd())
+            .wrap_err("failed to register LoLA file descriptor in epoll")?;
+        add_to_epoll(epoll_fd, hula.as_raw_fd())
+            .wrap_err("failed to register hula file descriptor in epoll")?;
+
+        Ok(Self {
+            lola,
+            hula,
+            epoll_fd,
+        })
+    }
+
+    pub fn run(mut self) -> Result<()> {
+        let proxy_start = Instant::now();
+        let mut connections = HashMap::new();
+        let mut events = [Event::new(Events::empty(), 0); 16];
+        let mut writer = BufWriter::with_capacity(786, self.lola.try_clone()?);
+        let mut battery = None;
+
+        debug!("Entering epoll loop...");
+        loop {
+            let epoll_timeout = -1;
+            let num_events = epoll::wait(self.epoll_fd, epoll_timeout, &mut events)
+                .wrap_err("failed to wait for epoll")?;
+            for event in &events[0..num_events] {
+                let notified_fd = event.data as i32;
+                if notified_fd == self.lola.as_raw_fd() {
+                    handle_lola_event(&mut self.lola, &mut connections, proxy_start, &mut battery)?;
+                } else if notified_fd == self.hula.as_raw_fd() {
+                    register_connection(&mut self.hula, &mut connections, self.epoll_fd)?;
+                } else {
+                    handle_connection_event(&mut connections, notified_fd, &mut writer, &battery)?;
+                }
+            }
+
+            if !connections
+                .values()
+                .filter_map(|connection| connection.last_time_received)
+                .any(|instant| instant.elapsed() < HULA_TIMEOUT)
+            {
```

Comment:
You changed this condition from `connections.is_empty()` to this one. I'm not happy with that change. IMHO, if at least there is one connection, this connection should have exclusive access to LOLA without HULA injecting control messages. Please change to the old predicate.

Reply:
No, I disagree. In the future, aliveness servers will attach without ever sending any data

Reply:
I agree, that this is not pretty, if you pause the framework (e.g. for debugging purposes). Do you have ideas?

Reply:
Okay, I understand your point but we need a compromise because both our points are important.

Reply:
What do you think of making the hula endpoints unidirectional? one for sending and one for receiving data?

Reply:
Hmmm, it would be one solution but I don't like that we would have another socket endpoint. What I could imagine is that all connected clients are read-only clients and are not counted. Once they send their first control message (e.g. our framework but not aliveness) they are opting in as being read-write and would be counted.

Reply:
That sounds nice! I'll try to implement that.

Reply:
Done, thanks for this proposal ;)

Reply:
Do we want to move `connection.is_sending_control_frames = true;` earlier s.t. it does not depend on successful LoLA socket writing? I would count a connection as `is_sending_control_frames` if we receive one but not if we successfully delivered it to LoLA. :thinking:

Reply:
I think that barely matters. This error case is hopefully never triggered. We should not invest to much time on that issue